git/archive.go

```diff
@@ -0,0 +1,274 @@
+package git
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/flate"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+)
+
+type ArchiveOptions struct {
+	Verbose            bool
+	List               bool
+	WorktreeAttributes bool
+	ArchiveFormat      string
+	BasePrefix         string
+	OutputFile         string
+	CompressionLevel   int
+}
+
+type ArchiveFormat int
+
+const (
+	archiveUnknow = ArchiveFormat(iota)
+	archiveTar
+	archiveTarGzip
+	archiveZip
+)
+
+var supportedArchiveFormats = map[string]ArchiveFormat{
+	"tar":    archiveTar,
+	"tgz":    archiveTarGzip,
+	"tar.gz": archiveTarGzip,
+	"zip":    archiveZip,
+}
+
+func findOutputFileFormat(output string) ArchiveFormat {
+	// if the output is empty return the default value, a tarball.
+	if output == "" {
+		return archiveTar
+	}
+
+	for k, v := range supportedArchiveFormats {
+		if strings.HasSuffix(strings.ToLower(output), k) {
+			return v
+		}
+	}
+	// The archive format is not found, return unknow.
+	return archiveUnknow
+}
+
+func createTarArchive(c *Client, opts ArchiveOptions, tgz bool, sha Sha1, entries []*IndexEntry) error {
+	fileOutput := os.Stdout
+
+	// If the output file is not empty use it instead of stdout
+	if opts.OutputFile != "" {
+		if file, err := os.Create(opts.OutputFile); err == nil {
+			fileOutput = file
+			defer file.Close()
+		} else {
+			return err
+		}
+	}
+
+	mtime := time.Now()
+
+	// If the sha is a tree we set the file modification time to the current time
+	// otherwise we must use the commit time.
+	if sha.Type(c) != "tree" {
+		if t, err := CommitID(sha).GetDate(c); err == nil {
+			mtime = t
+		}
+	}
+
+	var tw *tar.Writer
+
+	//
```

Comment:
forgot to delete this or write something?

Reply:
It's declared here because to use gzip compression the tar.Writer output must be set to a gzip.Writer.
See [this code](https://github.com/driusan/dgit/blob/4ccf41a07eedc0902e948cf6f607489e6849cdb6/git/archive.go#L82)

Reply:
This can be changed to the code below, but in my opinion the defer thing is a mess.
```
var fileOutput io.Writer = os.Stdout
...
// If gzip compression is enabled
if tgz {
	gw := gzip.NewWriter(fileOutput)
	fileOutput = gw
	defer gw.Close()
}

// Create the tar writer
tw := tar.NewWriter(fileOutput)
defer tw.Close()
```

Reply:
No, I just mean the empty "//" delimiter, not the code

Reply:
I just forgot to remove the empty comment :-)